handlers/secrets.go

```diff
@@ -28,30 +29,67 @@ const (
 
 // MakeSecretHandler makes a handler for Create/List/Delete/Update of
 //secrets in the Kubernetes API
-func MakeSecretHandler(namespace string, kube kubernetes.Interface) http.HandlerFunc {
+func MakeSecretHandler(defaultNamespace string, kube kubernetes.Interface) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		if r.Body != nil {
 			defer r.Body.Close()
 		}
 
-		client := kube.CoreV1().Secrets(namespace)
+		lookupNamespace, err, errCode := getLookupNamespace(defaultNamespace, r)
+		if err != nil {
+			http.Error(w, err.Error(), errCode)
+		}
+
+		client := kube.CoreV1().Secrets(lookupNamespace)
 
 		switch r.Method {
 		case http.MethodGet:
 			getSecretsHandler(client, w, r)
 		case http.MethodPost:
-			createSecretHandler(client, namespace, w, r)
+			createSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodPut:
-			replaceSecretHandler(client, namespace, w, r)
+			replaceSecretHandler(client, lookupNamespace, w, r)
 		case http.MethodDelete:
-			deleteSecretHandler(client, namespace, w, r)
+			deleteSecretHandler(client, lookupNamespace, w, r)
 		default:
 			w.WriteHeader(http.StatusBadRequest)
 			return
 		}
 	}
 }
 
+// getLookupNamespace returns the namespace specified in any request type
+func getLookupNamespace(defaultNamespace string, r *http.Request) (string, error, int) {
+	req := struct{ Namespace string }{Namespace: defaultNamespace}
+
+	switch r.Method {
+	case http.MethodGet:
+		q := r.URL.Query()
+		namespace := q.Get("namespace")
+
+		if len(namespace) > 0 {
+			req.Namespace = namespace
+		}
+
+	case http.MethodPost, http.MethodPut, http.MethodDelete:
+		body, _ := ioutil.ReadAll(r.Body)
+		err := json.Unmarshal(body, &req)
+		if err != nil {
+			log.Printf("error while getting namespace: %s\n", err)
+			return "", fmt.Errorf("unable to unmarshal Secret request"), http.StatusBadRequest
+		}
+
+		// Reconstruct Body
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	}
+
+	if req.Namespace == "kube-system" {
```

Comment:
I think this is yet another issue. We have checks for the `kube-system` namespace in various place, would need a slightly bigger refactor.

Reply:
I am not sure I agree. This change allows the user to modify secrets in namespaces that it isn't allowed to deploy functions to. There is already code for getting and checking the list of allowed. The refactor here would be to call the `list` method and then, in the `if` block, use the `findNamepsace` method. In fact, you could combine those into a single statement, so it could even be a one line change to the PR.

Refactoring the other checks can definitely be done in a separate PR, but I don't think we should ignore it here especially when there is a drop in solution already implemented.

But, I am ok with merging this if @alexellis wants to get the functionality in first and refactor later

Reply:
Oh yes, I meant opening a PR for the other areas, I'll quickly add the check for this function

Reply:
Awesome and thanks!